### PR TITLE
Use the right interpreter and kernel with notebooks

### PR DESCRIPTION
**Pull request by DonJayamanne**
For #5489 

This has two fixes
* Ensure we select the right interpreter associated with a kernelspec
* Ensure we pick the right kernelspec based on notebook metadata

### CODE REVIEW

File: src/client/datascience/kernel-launcher/localKernelFinder.ts [hunk 1]
```diff
@@ -276,62 +278,63 @@ export class LocalKernelFinder implements ILocalKernelFinder {
...
-                    `Kernel ${kernelSpec.name} is not python based so does not have an interpreter.`
-                );
-                return false;
-            }
-
-            // 1. Check if current interpreter has the same path
+    ): PythonEnvironment | undefined {
+        // If we know for a fact that the kernel spec is a Non-Python kernel, then return nothing.
+        if (kernelSpec.language && kernelSpec.language !== PYTHON_LANGUAGE) {
+            return;
```
Comment (rchiodo) [inline src/client/datascience/kernel-launcher/localKernelFinder.ts:287]: Why remove the tracing?

**Review by rchiodo — APPROVED** [review]

File: src/client/datascience/kernel-launcher/localKernelFinder.ts (hunk 1 above)
Comment (DonJayamanne) [reply src/client/datascience/kernel-launcher/localKernelFinder.ts:287]: I'll add that back.

File: src/client/datascience/kernel-launcher/localKernelFinder.ts [hunk 2]
```diff
@@ -193,18 +193,20 @@ export class LocalKernelFinder implements ILocalKernelFinder {
...
-                        id: getKernelId(k, matchingInterpreters[0])
+                        interpreter: matchingInterpreter,
+                        id: getKernelId(k, matchingInterpreter)
                     };
 
                     // If interpreters were found, remove them from the interpreter list we'll eventually
                     // return as interpreter only items
-                    filteredInterpreters = filteredInterpreters.filter((i) => !matchingInterpreters.includes(i));
+                    filteredInterpreters = filteredInterpreters.filter(
+                        (i) => matchingInterpreter === i || this.fs.areLocalPathsSame(matchingInterpreter.path, i.path)
```
Comment (IanMatthewHuff) [inline src/client/datascience/kernel-launcher/localKernelFinder.ts:208]: This local path check doesn't look right to me. We have PythonEnvironments here, right? As far as I know (not the expert here) you can't match those by path. With conda without specifying a specific python multiple conda envs can share one local interpreter path.

**Review by IanMatthewHuff — CHANGES_REQUESTED** [review]
Sorry, maybe I'm off on this (and this looks like it was an existing issue as well) but it looks to me like there is an issue here with using interpreter paths to match, when I believe those can be shared by conda python environments at least.

File: src/client/datascience/kernel-launcher/localKernelFinder.ts (hunk 2 above)
Comment (DonJayamanne) [reply src/client/datascience/kernel-launcher/localKernelFinder.ts]: Fixed

**Review by IanMatthewHuff — APPROVED** [review]
Approving, though I would mention that it looks like that local path checking might be an issue in the existing code as well? Not fully sure from a quick review, but I think that we might be making bad assumptions for conda here.

**Comment by DonJayamanne** [comment]
> ough I would mention that it looks like that local path checking might be an issue in the 

I removed that change.

**Comment by IanMatthewHuff** [comment]
> > ough I would mention that it looks like that local path checking might be an issue in the
> 
> I removed that change.

@DonJayamanne Sorry, maybe I wasn't clear. I don't mean in your change. I mean the existing code that didn't change seems to also be looking at localpath in a way that could cause issues for conda environments.

**Comment by DonJayamanne** [comment]
@IanMatthewHuff Please could you create an issue for that so we can discuss that, could definitely be a problem for users

**Comment by IanMatthewHuff** [comment]
> @IanMatthewHuff Please could you create an issue for that so we can discuss that, could definitely be a problem for users

Added #5650

**Pull request merged**